waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/server/FederatedHMSHandlerHive3.java

```diff
@@ -0,0 +1,674 @@
+    public GetCatalogsResponse get_catalogs()
+            throws MetaException, TException
+    {
+        return getPrimaryClient().get_catalogs();
+    }
+
+    @Override
+    @Loggable(value = Loggable.DEBUG, skipResult = true, name = INVOCATION_LOG_NAME)
+    public void drop_catalog(DropCatalogRequest dropCatalogRequest)
+            throws NoSuchObjectException, InvalidOperationException, MetaException, TException
+    {
+        getPrimaryClient().drop_catalog(dropCatalogRequest);
+    }
+
+    @Override
+    @Loggable(value = Loggable.DEBUG, skipResult = true, name = INVOCATION_LOG_NAME)
+    public void add_unique_constraint(AddUniqueConstraintRequest addUniqueConstraintRequest)
+            throws NoSuchObjectException, MetaException, TException
+    {
+        getPrimaryClient().add_unique_constraint(addUniqueConstraintRequest);
+    }
+
+    @Override
+    @Loggable(value = Loggable.DEBUG, skipResult = true, name = INVOCATION_LOG_NAME)
+    public void add_not_null_constraint(AddNotNullConstraintRequest addNotNullConstraintRequest)
+            throws NoSuchObjectException, MetaException, TException
+    {
+        getPrimaryClient().add_not_null_constraint(addNotNullConstraintRequest);
+    }
+
+    @Override
+    @Loggable(value = Loggable.DEBUG, skipResult = true, name = INVOCATION_LOG_NAME)
+    public void add_default_constraint(AddDefaultConstraintRequest addDefaultConstraintRequest)
+            throws NoSuchObjectException, MetaException, TException
+    {
+        getPrimaryClient().add_default_constraint(addDefaultConstraintRequest);
+    }
+
+    @Override
+    @Loggable(value = Loggable.DEBUG, skipResult = true, name = INVOCATION_LOG_NAME)
+    public void add_check_constraint(AddCheckConstraintRequest addCheckConstraintRequest)
+            throws NoSuchObjectException, MetaException, TException
+    {
+        getPrimaryClient().add_check_constraint(addCheckConstraintRequest);
+    }
+
+    @Override
+    @Loggable(value = Loggable.DEBUG, skipResult = true, name = INVOCATION_LOG_NAME)
+    public void truncate_table(String dbName, String tableName, List<String> partNames)
+            throws MetaException, TException
+    {
+        getPrimaryClient().truncate_table(dbName, tableName, partNames);
+    }
+
+    @Override
+    @Loggable(value = Loggable.DEBUG, skipResult = true, name = INVOCATION_LOG_NAME)
+    public List<String> get_materialized_views_for_rewriting(String view)
+            throws MetaException, TException
+    {
+        return getPrimaryClient().get_materialized_views_for_rewriting(view);
+    }
+
+    @Override
+    @Loggable(value = Loggable.DEBUG, skipResult = true, name = INVOCATION_LOG_NAME)
+    public Materialization get_materialization_invalidation_info(CreationMetadata creationMetadata, String s)
+            throws MetaException, InvalidOperationException, UnknownDBException, TException
+    {
+        return null;
```

Comment:
So, for these methods which are currently returning `null` I think they should either be implemented or they could throw an `UnsupportedOperationException` for now until someone actually needs them and implements them. I take it you haven't needed them for your use case?

Reply:
Generally any API call that does not contain a database cannot be correctly proxied. We normally just  called the "primary" metastore for that. Reasoning is that most users use the primary for reading and writing and proxy to the federating ones just for reading which "generally" didn't involve these calls. 

That way at least users can do all the work on the primary via the proxy.

Any approach will have limitations but that has served us well in the past, so perhaps instead of throwing exception or returning null we delegate to the primary metastore.

There might be ones where it really doesn't make sense though and throwing an exception is a good alternative.